solidity/test/submitERC721LogicCall.ts

```diff
@@ -0,0 +1,448 @@
+    powers,
+    valsetNonce: currentValsetNonce,
+    rewardAmount: 0,
+    rewardToken: ZeroAddress
+  }
+
+  let logicCallSubmitResult = await gravity.submitLogicCall(
+    valset,
+    sigs,
+    logicCallArgs
+  );
+
+  //check ownership of ERC721 tokens now transferred to signers
+  for (let i = 0; i < numTxs; i++) {
+    expect(
+      (await testERC721.ownerOf(tokenIds[i]))
+    ).to.equal(signers[i + 5].address);
+  }
+
+  // check that the relayer was paid
+  expect(
+    await (
+      await testERC20.functions.balanceOf(await logicCallSubmitResult.from)
+    )[0].toNumber()
+  ).to.equal(9010);
+
+  expect(
+    (await testERC20.functions.balanceOf(await signers[20].getAddress()))[0].toNumber()
+  ).to.equal(0);
+
+  expect(
+    (await testERC20.functions.balanceOf(gravity.address))[0].toNumber()
+  ).to.equal(990);
+
+  expect(
+    (await testERC20.functions.balanceOf(ERC721LogicContract.address))[0].toNumber()
+  ).to.equal(0);
+
+  expect(
+    (await testERC20.functions.balanceOf(await signers[0].getAddress()))[0].toNumber()
+  ).to.equal(9010);
+}
+
+describe("submitLogicCall tests", function () {
+  it.only("throws on malformed current valset", async function () {
+    await expect(runTest({ malformedCurrentValset: true })).to.be.revertedWith(
+      "MalformedCurrentValidatorSet()"
+    );
+  });
+
+  it.only("throws on wrong ownerOf", async function () {
+    await expect(runTest({ wrongOwner: true })).to.be.revertedWith(
+      "Ownable: caller is not the owner"
+    );
+  });
+
+  it.only("throws on invalidation nonce not incremented", async function () {
+    await expect(runTest({ invalidationNonceNotHigher: true })).to.be.revertedWith(
+      "InvalidLogicCallNonce(0, 0)"
+    );
+  });
+
+  it.only("throws on non matching checkpoint for current valset", async function () {
+    await expect(
+      runTest({ nonMatchingCurrentValset: true })
+    ).to.be.revertedWith(
+      "IncorrectCheckpoint()"
+    );
+  });
+
+  it.only("throws on bad validator sig", async function () {
+    await expect(runTest({ badValidatorSig: true })).to.be.revertedWith(
+      "InvalidSignature()"
+    );
+  });
+
+  it.only("allows zeroed sig", async function () {
+    await runTest({ zeroedValidatorSig: true });
+  });
+
+  it.only("throws on not enough signatures", async function () {
+    await expect(runTest({ notEnoughPower: true })).to.be.revertedWith(
+      "InsufficientPower(2807621889, 2863311530)"
+    );
+  });
+
+  it.only("does not throw on barely enough signatures", async function () {
+    await runTest({ barelyEnoughPower: true });
+  });
+
+  it.only("throws on timeout", async function () {
+    await expect(runTest({ timedOut: true })).to.be.revertedWith(
+      "LogicCallTimedOut()"
+    );
+  });
+
+});
+
+// This test produces a hash for the contract which should match what is being used in the Go unit tests. It's here for
+// the use of anyone updating the Go tests.
+describe("logicCall Go test hash", function () {
```

Comment:
this can be removed, no need to duplicate it.

Reply:
Removed